### ISSUE TITLE
confused about loss function

### ISSUE THREAD

**Issue by helson73** [naver-ai/pcme#2]
Great work! But I am confused about some details in loss function.

[`-((logit * matched - torch.stack((logit, -logit), dim=2).logsumexp(dim=2, keepdim=False)).logsumexp(dim=1)) + np.log(logit.size(1))`](https://github.com/naver-ai/pcme/blob/587aa0de710f3f40eac42cd97657c1a9dcfc6ebc/criterions/probemb.py#L86)
I guess this part is equivalent to y = 1 / (1+exp(-2x)), which is compressed sigmoid, not sigmoid itself.
Although this will not affect performance, but may report wrong learned parameters (a, b).

In addition, reduction is always [done in 'sum' ](https://github.com/naver-ai/pcme/blob/587aa0de710f3f40eac42cd97657c1a9dcfc6ebc/criterions/probemb.py#L201) whatever [self.reduction](https://github.com/naver-ai/pcme/blob/587aa0de710f3f40eac42cd97657c1a9dcfc6ebc/criterions/probemb.py#L125) is set.

**Comment by SanghyukChun** [comment]
Hi, I cannot fully understand your question, but I presume that your question is "is the following loss function is a sigmoid function?"
https://github.com/naver-ai/pcme/blob/587aa0de710f3f40eac42cd97657c1a9dcfc6ebc/criterions/probemb.py#L86-L89

No. This is not a sigmoid function. It is the MC-soft contrastive loss (negative log-likelihood) by the matching probability. Sigmoid function and learnable affine transform (`a`, `b`) is used for the matching probability function.

Please read the following slides (from 23-26) for a deeper understanding
https://docs.google.com/presentation/d/1dGQUqud3iMld-UgMMlRQuqA7JhzndMfzFoKaFAwZ58I/edit#slide=id.ge3b8390885_0_32

You can find the exactly same information in the code as well
https://github.com/naver-ai/pcme/blob/587aa0de710f3f40eac42cd97657c1a9dcfc6ebc/criterions/probemb.py#L51-L73

The `reduction` argument is not used in this code but is used for compatibility with other Torch loss functions.

**Comment by helson73** [comment]
Actually what I mean is ..
![캡처](https://user-images.githubusercontent.com/4516503/125770017-5cf4cc5b-9262-49da-96ce-f4b40cc4562f.JPG)
But never mind, this has no affect to learning any way.

Thank you very much for sharing this code. Your work helped me a lot.

**Comment by SanghyukChun** [comment]
@helson73 
Yes, as denoted in the code (as well as the attached slides), we use `exp(x) / (exp(x) + exp(-x))` as a sigmoid function. Since we have the learnable affine transform (`a` and `b`), it actually does not affect the training procedure and the final performances.

**Comment by helson73** [comment]
I have a few more questions, if you don't mind I post here ...
(it's bad I didn't attend your presentation in CVPR ...)

I applied your method in anther somehow similar but much more different task,
But I frequently got -inf or nan loss (maybe caused by gradient overflow or zero division), even the training accidentally went through this, learning was too slow that results never reach the standard triplet loss.

And my questions are:
1) Two stage training. You applied two-stage procedure, former phase for warming up and latter for fine-tuning. Is this mandatory for training (in other word training will not succeed without two-stage training) or just for speed up convergence ?

2) KL divergence. When KL divergence term is involved, the training objective is equivalent to maximizing variational lower bound of information bottleneck (VIB).  ​Have you tried to derive any other variants VIB rather than contrastive loss ? I am wondering if there is alternative way like directly optimizing triplet-like objective in variantional bayes fashion. [This paper](https://arxiv.org/pdf/1506.05011.pdf) may help but it's a generative model. 

3) Practices in stability of training. I guess the reasons I frequently got overflowed gradients could be: a) numerical stability issue, which is inevitable at early development stage in every neural networks related task, especially when lots of exponential, logarithm and division operations are involved. b) stochastic collapse issue, wrong hyper-parameters, large value of co-variance, ... and so on. If you don't mind, could you share some know-hows you learned in your early development stage ? I also encountered lots of similar situations in various tasks, but this one, involves too many factors, any advise could be helpful.

Thanks again.

**Comment by SanghyukChun** [comment]
No problem. 
Disclaimer: I don't know your application, backbones, datasets, so my answers may not solve your question.
Note that the MC-soft contrastive loss is reduced to the soft plus operation, which is known to be very stable. I didn't face the inf or nan loss when I tested my code on MS-COCO and CUB. Our method is quite stable, and the only unstable training procedure I faced was the out-of-memory error (because we need to compute 7 x 7 more pairs than the non-sampling counterparts)

1. "Warm-up is mandatory?"
    - My implementation is based on [VSE++](https://github.com/fartashf/vsepp) and [PVSE](https://github.com/yalesong/pvse). These methods employ the warm-up, and I followed the previous practice. I tested "no warmup" before, and I got slightly worse results than the warmup-ed one. But, the warmup is for better performance (or speed up convergence, as you mentioned), not the mandatory process for stable training.
2. VIB for triplet loss
    - I am not sure that I did correctly understand your question, but I presume that your question is, "is it possible to apply VIB (KL divergence in my code) loss to triplet?". The answer is yes. [View-Invariant Probabilistic Embedding for Human Pose (ECCV 2020)](https://www.ecva.net/papers/eccv_2020/papers_ECCV/papers/123500052.pdf) applied the probabilistic embedding to triplet loss (with some modification), and they also applied KL divergence to the triplet loss. If this is not the answer you want, please clarify your question again.
3. Recipes for stable training
    - Disclaimer: I never faced the unstable loss explosion when I work with probabilistic embedding. These answers are just my guesses, may not work for your application.
    - (a) as derived in this issue, the actual implementation is based on the log sum exp operation (or soft-plus), which is numerically very stable. If you have implemented the loss function directly from the sigmoid, you can face the numerical issue. In this case, you have to re-implement your code with log sum exp operations as my code.
    - (b) stochastic collapse issue is rarely happening if you use log sum exp implementation. However, if you implement triplet loss with probabilistic distances (e.g., KL divergence, W-distance), I have faced a lot of loss explosion. My advice is to use my implementation, not the manually designed triplet + probability distances.
    - (b) hyperparameters can be matter. One typically faced issue is the initial value of `a` and `b`. I omitted the details in the paper, but the initial `a` and `b` values could be important to your model (in terms of convergence). Note that `a` and `b` determine the relative margin value of the soft contrastive loss and the gradient of the loss function as well. My advice is to use large `a` and `b`, such as `10`. In my implementation, I used `15` from the cross-validation hyperparameter search at the CUB caption dataset.
    - (c) variance network can be matter as well. For example, if you have sigmoid (or any other normalization layers) at the output of the variance network, your network will have very limited power to represent uncertainty, which can lead to very bad performances or unstable convergence. Also, it could be important to check the initial values produced by the uncertainty network, e.g., it could be almost zero or very large due to the network architecture.

**Comment by SanghyukChun** [comment]
I attached the full derivation for MC contrastive loss again (for better understanding)
> Please read the following slides (from 23-26) for a deeper understanding
> https://docs.google.com/presentation/d/1dGQUqud3iMld-UgMMlRQuqA7JhzndMfzFoKaFAwZ58I/edit#slide=id.ge3b8390885_0_32

<img width="1291" alt="Screen Shot 2021-07-17 at 4 48 31 PM" src="https://user-images.githubusercontent.com/4919315/126030179-46e841c5-f9c0-4100-b9af-901ec618958d.png">
<img width="1119" alt="Screen Shot 2021-07-17 at 4 48 37 PM" src="https://user-images.githubusercontent.com/4919315/126030178-94dd5056-1123-422c-a8f0-a21481d3fd01.png">
<img width="1304" alt="Screen Shot 2021-07-17 at 4 48 44 PM" src="https://user-images.githubusercontent.com/4919315/126030177-af6bf19a-cb7c-4b84-9a31-6b0868464e4d.png">
<img width="1094" alt="Screen Shot 2021-07-17 at 4 48 50 PM" src="https://user-images.githubusercontent.com/4919315/126030171-26f5f706-75ec-4041-b536-28f7f2530ebd.png">

**Comment by SanghyukChun** [comment]
@helson73 
Closing the issue, assuming the answer resolves the problem.
Please re-open the issue or open another issue as necessary.

**Comment by helson73** [comment]
> @helson73
> Closing the issue, assuming the answer resolves the problem.
> Please re-open the issue or open another issue as necessary.

Hi, thanks for your answers and sorry for the late reply.
I am now revisiting my overall training process (thanks to your help) and I will update and reorganize my latest learned insights and questions (if any) as soon as possible.
